data-import/desktop-csv-import/desktop-csv-import.adoc

```diff
@@ -0,0 +1,420 @@
+
+image::{img}desktop_import_data_model.jpg[link="{img}desktop_import_data_model.jpg",role="popup-link"]
+
+We have two nodes - one for a product and one for an order.
+Each of those nodes have properties that we have from our CSV files.
+For the `Product`, we have ID, name, and unit cost.
+For the `Order`, we have ID, date/time, and country where it is going.
+
+The order-details.csv file defines the relationship between those two nodes.
+That file has the product ID, the order ID it belongs to, and the quantity of the product on the order.
+So, in our data model, this becomes our `CONTAINS` relationship between `Product` and `Order` nodes.
+We also include a property of `quantityOrdered` on the relationship because the product quantity value only exists when a product is related to an order.
+
+Now that we know the types of nodes and relationships we will have and the properties involved, we can construct the Cypher statements to create the data for this model.
+
+=== Avoiding duplicates and increasing performance
+
+One final thing we need to think about before we create data in the graph is ensuring values are unique and performance is efficient.
+To handle this, we can use constraints.
+Just as with other databases, constraints ensure data integrity criteria is not violated, while simultaneously indexing the property with the constraint for faster query performance.
+
+In this exercise, we will add two constraints - one for `productId` and one for `orderId`.
+This will ensure when we create a new node of each of those types or a relationship to connect them, we know the entities are unique and indexed.
+
+Good practice is to add any constraints and indexes before loading to avoid existing violations and ensure data is added to the index as it is created.
+Below is the Cypher for adding indexes.
+
+[source, cypher]
+----
+CREATE CONSTRAINT ON (p:Product) ASSERT p.id IS UNIQUE;
+CREATE CONSTRAINT ON (o:Order) ASSERT o.id IS UNIQUE;
+----
+
+=== Cypher
+
+Now we are ready to create the Cypher for creating the data in the graph!
+We will use `CREATE` statements where we are sure that we will not have duplicate rows in our CSV file.
+For values we know exist in the database already, we will first use `MATCH` to find them before executing any updates or inserts.
+
+.Products
+To start, let us load the products into the graph.
+We start with our `LOAD CSV` statement from above, then we add the Cypher to create the data from the CSV into our graph model.
+We can use the `CREATE` statement, as we feel confident that we have our constraint in place, no duplicate rows in our file, and no existing data in our graph.
+The properties will be set to the converted values we handled earlier in this guide.
+
+[source, cypher]
+----
+LOAD CSV FROM 'file:///desktop-csv-import/products.csv' AS row
+CREATE (p:Product {productId: toInteger(row[0])})
```

Comment:
I don't think this is good practice, naming the columns early and using the names later makes for more maintainable queries.

Reply:
Sorry about that. I'll fix it!